### PR TITLE
feat: Adding Tenant Information to Redis Cache for quick fetch

### LINKED ISSUE #33083: [Task]: Implement Redis caching for tenant configuration
### Description

The tenants/current api is the slowest component of consolidated/view api. Reducing the latency of tenants/current api will reduce the latency of backend in view experience for all users.

Relevant Slack thread : https://theappsmith.slack.com/archives/C024GUDM0LT/p1713933688987279

In the diagram in the slack thread, the blue line just below the magenta line at the top is the contribution of tenants/current to consolidated api. The diagram shows that tenants/current is 90% contributor to consolidated view api.

Upon debugging the cause of latency from tenants/current, it is found that the api makes a DB call to fetch the tenant configuration every time. 

Caching the tenant config ID is already present in the codebase. Extending that approach, we can cache the whole tenant config and update it when a user changes admin settings.

Not making the DB call will reduce the latency of tenants/current api and decrease the latency of consolidated view api as well.

CE PR: https://github.com/appsmithorg/appsmith/pull/33641
Counterpart EE PR: https://github.com/appsmithorg/appsmith-ee/pull/4275

### Subtasks
- Implement caching for fetching tenant configuration
- Invalidate cache if tenant configuration gets updated, update the database and rebuild cache.

### PR DESCRIPTION
**Pull request by NilanshBansal**
## Description

The tenant is fetched multiple times across the appsmith codebase but is rarely updated (from the admin settings). Every time a fetch call to the database is costly both in terms of resources and time taken. 
The consolidated api also makes a call to fetch the tenant and return to the client. To improve the performance of fetching the tenant information, we are moving the tenant information to redis cache for quicker fetch. 
This will improve the performance of the consolidated api and also reduce the time taken by all the different functionalities within the backend codebase which depend on tenant to process further. 

> The old PR implementation https://github.com/appsmithorg/appsmith/pull/33309 had to be reverted due to the tenant GAC permissions not in sync between database and redis. RCA [ref](https://www.notion.so/appsmith/Reversion-for-Tenant-Caching-implementation-after-merging-to-release-cd720b9959e4413f98decb884c375ad2).
This PR uses the same branch as the old PR and builds the pending functionalities to complete the implementation. 

**Counterpart EE PR**: https://github.com/appsmithorg/appsmith-ee/pull/4275

**TL;DR**
Adds tenant information `tenantService.getDefaultTenant()` to redis.

Fixes #33083, #33504, https://github.com/appsmithorg/appsmith/issues/33578
## Automation

/ok-to-test tags="@tag.Settings"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/9253953003>
> Commit: 7b4bf8d6b0b6273988c71ff37b615f81e36f03f4
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=9253953003&attempt=1" target="_blank">Click here!</a>

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by NilanshBansal** [comment]
@abhvsn @sharat87 can you please review [this](https://github.com/appsmithorg/appsmith/pull/33641/commits/d11d404f3c06109859282601d7bfed4f8feb1922) and [this](https://github.com/appsmithorg/appsmith/pull/33641/commits/0d22cd05203b50301d2354d88ca69d57219decaa) commits which are recently merged in. Other commits are older which originated from reopening the old merged PR.

These commits resolve the issues https://github.com/appsmithorg/appsmith/issues/33504 and https://github.com/appsmithorg/appsmith/issues/33578

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CacheableRepositoryHelperCEImpl.java [hunk 1]
```diff
@@ -166,4 +167,33 @@ public Mono<String> getInstanceAdminPermissionGroupId() {
...
+            }
+            return tenant;
+        });
+    }
+
+    @CacheEvict(cacheName = "tenant", key = "{#tenantId}")
+    @Override
+    public Mono<Void> evictCachedTenant(String tenantId) {
+        return Mono.empty().then();
+    }
```
[the remark this answers is not in the document]
Comment (NilanshBansal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CacheableRepositoryHelperCEImpl.java:198]: This is already handled in the existing test cases, and the newly added test

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/TenantRepositoryCE.java [hunk 2]
```diff
@@ -5,6 +5,7 @@
 import reactor.core.publisher.Mono;
 
 public interface TenantRepositoryCE extends BaseRepository<Tenant, String>, CustomTenantRepositoryCE {
-
+    // Use tenantService.getDefaultTenant() instead of this method as it is cached to redis.
+    @Deprecated(forRemoval = true)
```
[the remark this answers is not in the document]
Comment (NilanshBansal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/TenantRepositoryCE.java:9]: Purposely used to fetch data from db and not cache.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CacheableRepositoryHelperCE.java [hunk 3]
```diff
@@ -18,4 +19,8 @@ public interface CacheableRepositoryHelperCE {
     Mono<String> getDefaultTenantId();
 
     Mono<String> getInstanceAdminPermissionGroupId();
+
+    Mono<Tenant> fetchDefaultTenant(String tenantId);
+
+    Mono<Void> evictCachedTenant(String tenantId);
```
[the remark this answers is not in the document]
Comment (NilanshBansal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CacheableRepositoryHelperCE.java:25]: Not required as these are getting covered with the existing and new tests.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Tenant.java [hunk 4]
```diff
@@ -10,13 +10,15 @@
 import org.springframework.data.annotation.Transient;
 import org.springframework.data.mongodb.core.mapping.Document;
 
+import java.io.Serializable;
+
```
[the remark this answers is not in the document]
Comment (NilanshBansal) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Tenant.java:14]: Not required as PricingPlan is an enum.
cc: @abhvsn

**Comment by NilanshBansal** [comment]
All tests have passed earlier before [this commit](https://github.com/appsmithorg/appsmith/pull/33641/commits/7b4bf8d6b0b6273988c71ff37b615f81e36f03f4).  Running only a subset of tests now as the change is only done to a server unit test file.
image

**Review by abhvsn — APPROVED** [review]
